src/simplegmres.jl

```diff
@@ -0,0 +1,600 @@
+    else
+        w .= b
+    end
+    PlisI || ldiv!(r₀, Pl, w)  # r₀ = Pl(b - Ax₀)
+    β = _norm2(r₀)         # β = ‖r₀‖₂
+
+    rNorm = β
+    ε = abstol + reltol * rNorm
+
+    return SimpleGMRESCache{true}(memory, n, restart, maxiters, blocksize, ε, PlisI, PrisI,
+        Pl, Pr, Δx, q, p, x, A, b, abstol, reltol, w, V, s, c, z, R, β, warm_start)
+end
+
+function SciMLBase.solve!(cache::SimpleGMRESCache{true}, lincache::LinearCache)
+    @unpack memory, n, restart, maxiters, blocksize, ε, PlisI, PrisI, Pl, Pr = cache
+    @unpack Δx, q, p, x, A, b, abstol, reltol, w, V, s, c, z, R, β, warm_start = cache
+    bsize = n ÷ blocksize
+
+    __batch = Base.Fix2(reshape, (blocksize, bsize))
+
+    T = eltype(x)
+    q = PlisI ? w : q
+    r₀ = PlisI ? w : q
+    xr = restart ? Δx : x
+
+    if β == 0
+        return SciMLBase.build_linear_solution(lincache.alg, x, r₀, lincache;
+            retcode = ReturnCode.Success)
+    end
+
+    rNorm = β
+    npass = 0        # Number of pass
+
+    iter = 0        # Cumulative number of iterations
+    inner_iter = 0  # Number of iterations in a pass
+
+    # Tolerance for breakdown detection.
+    btol = eps(T)^(3 / 4)
+
+    # Stopping criterion
+    breakdown = false
+    inconsistent = false
+    solved = rNorm ≤ ε
+    inner_maxiters = maxiters
+    tired = iter ≥ maxiters
+    inner_tired = inner_iter ≥ inner_maxiters
+    status = ReturnCode.Default
+
+    while !(solved || tired || breakdown)
+        # Initialize workspace.
+        # TODO: Check that not zeroing out (V, s, c, R, z) doesn't lead to incorrect results.
+        nr = 0  # Number of coefficients stored in Rₖ.
+
+        if restart
+            xr .= zero(T)  # xr === Δx when restart is set to true
+            if npass ≥ 1
+                mul!(w, A, x)
+                axpby!(one(T), b, -one(T), w)
+                PlisI || ldiv!(r₀, Pl, w)
+            end
+        end
+
+        # Initial ζ₁ and V₁
+        β = _norm2(__batch(r₀), 1)
+        z[1] .= vec(β)
+        V[1] .= vec(__batch(r₀) ./ β)
+
+        npass = npass + 1
+        inner_iter = 0
+        inner_tired = false
+
+        while !(solved || inner_tired || breakdown)
+            # Update iteration index
+            inner_iter += 1
+            # Update workspace if more storage is required and restart is set to false
+            if !restart && (inner_iter > memory)
+                append!(R, [similar(first(R), bsize) for _ in 1:inner_iter])
+                push!(s, similar(first(s), bsize))
+                push!(c, similar(first(c), bsize))
+            end
+
+            # Continue the Arnoldi process.
+            p = PrisI ? V[inner_iter] : p
+            PrisI || ldiv!(p, Pr, V[inner_iter])  # p ← Nvₖ
+            mul!(w, A, p)                         # w ← ANvₖ
+            PlisI || ldiv!(q, Pl, w)                 # q ← MANvₖ
+            for i in 1:inner_iter
+                sum!(R[nr + i]', __batch(V[i]) .* __batch(q))
```

Comment:
@amontoison see this line. For the batched case (N x B input size), the `dot` needs to only reduce to `1 x B` instead of reducing to a scalar.